core/network-libp2p/src/custom_proto/behaviour.rs

```diff
@@ -0,0 +1,471 @@
+				peer_id: peer_id.clone(),
+				event: CustomProtosHandlerIn::EnablePassive,
+			});
+		}
+
+		self.enabled_peers.insert(peer_id, endpoint);
+	}
+
+	fn inject_disconnected(&mut self, peer_id: &PeerId, _: ConnectedPoint) {
+		while let Some(pos) = self.open_protocols.iter().position(|(p, _)| p == peer_id) {
+			let (_, protocol_id) = self.open_protocols.remove(pos);
+
+			let event = CustomProtosOut::CustomProtocolClosed {
+				protocol_id,
+				peer_id: peer_id.clone(),
+				result: Ok(()),
+			};
+
+			self.events.push(NetworkBehaviourAction::GenerateEvent(event));
+
+			// Trigger a `connect_to_nodes` round.
+			self.next_connect_to_nodes = Delay::new(Instant::now());
+		}
+
+		self.enabled_peers.remove(peer_id);
+	}
+
+	fn inject_node_event(
+		&mut self,
+		source: PeerId,
+		event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
+	) {
+		match event {
+			CustomProtosHandlerOut::CustomProtocolClosed { protocol_id, result } => {
+				let pos = self.open_protocols.iter().position(|(s, p)|
+					s == &source && p == &protocol_id
+				);
+
+				if let Some(pos) = pos {
+					self.open_protocols.remove(pos);
+				} else {
+					debug_assert!(false, "Couldn't find protocol in open_protocols");
+				}
+
+				let event = CustomProtosOut::CustomProtocolClosed {
+					protocol_id,
+					result,
+					peer_id: source,
+				};
+
+				self.events.push(NetworkBehaviourAction::GenerateEvent(event));
+			}
+			CustomProtosHandlerOut::CustomProtocolOpen { protocol_id, version } => {
+				debug_assert!(!self.open_protocols.iter().any(|(s, p)|
+					s == &source && p == &protocol_id
+				));
+				self.open_protocols.push((source.clone(), protocol_id));
+
+				if let Some(address) = self.enabled_peers.get(&source) {
+					let event = CustomProtosOut::CustomProtocolOpen {
+						protocol_id,
+						version,
+						peer_id: source,
+						endpoint: address.clone()
+					};
+
+					self.events.push(NetworkBehaviourAction::GenerateEvent(event));
+				}
+			}
+			CustomProtosHandlerOut::CustomMessage { protocol_id, data } => {
+				let event = CustomProtosOut::CustomMessage {
+					peer_id: source,
+					protocol_id,
+					data,
+				};
+
+				self.events.push(NetworkBehaviourAction::GenerateEvent(event));
+			}
+		}
+	}
+
+	fn poll(
+		&mut self,
+		params: &mut PollParameters<NetTopology>,
+	) -> Async<
+		NetworkBehaviourAction<
+			<Self::ProtocolsHandler as ProtocolsHandler>::InEvent,
+			Self::OutEvent,
+		>,
+	> {
+		loop {
+			match self.next_connect_to_nodes.poll() {
+				Ok(Async::Ready(())) => self.connect_to_nodes(params),
+				Ok(Async::NotReady) => break,
+				Err(err) => {
+					warn!(target: "sub-libp2p", "Connect-to-nodes timer errored: {:?}", err);
+					break
+				}
+			}
+		}
+
+		// Clean up `banned_peers`
+		self.banned_peers.retain(|(_, end)| *end < Instant::now());
+		self.banned_peers.shrink_to_fit();
+
+		if !self.events.is_empty() {
+			return Async::Ready(self.events.remove(0))
```

Comment:
Right now it's a `SmallVec`, and since the elements are quite small I think it shouldn't be very expensive to remove the first element.

I think we should eventually remove this events system, but that requires some big modifications of the `NetworkBehaviour` trait.